File: Source/Pegasus/Pegasus/Platform/Graphics/OpenGL3/VertexLayout.GL3.cs
namespace Pegasus.Platform.Graphics.OpenGL3
{
	using System;
	using Bindings;
	using Interface;
	using Utilities;

	/// <summary>
	///     Represents a description of the memory layout and other properties of the vertex data that is fed into the
	///     input-assembler stage of the graphics pipeline.
	/// </summary>
	internal unsafe class VertexLayoutGL3 : GraphicsObjectGL3, IVertexLayout
	{
		/// <summary>
		///     Initializes a new instance.
		/// </summary>
		/// <param name="graphicsDevice">The graphics device this instance belongs to.</param>
		/// <param name="description">A description of the vertex layout.</param>
		public VertexLayoutGL3(GraphicsDeviceGL3 graphicsDevice, ref VertexLayoutDescription description)
			: base(graphicsDevice)
		{
			//NativeObject = GLContext.Allocate(_gl.GenVertexArrays, "VertexLayout");
			//_gl.BindVertexArray(NativeObject);

			//if (description.IndexBuffer != null)
			//{
			//	var handle = ((BufferGL3)description.IndexBuffer.BufferObject).Handle;
			//	_gl.BindBuffer(GL.ElementArrayBuffer, handle);

			//	IndexOffset = description.IndexOffset;
			//	IndexType = description.IndexBuffer.IndexSize.Map();
			//	IndexSizeInBytes = description.IndexBuffer.IndexSize.GetIndexSizeInBytes();
			//}

			//for (var i = 0; i < description.Bindings.Length; ++i)
			//{
			//	var type = description.Bindings[i].Format.GetVertexDataType();
			//	var size = description.Bindings[i].Format.GetVertexDataComponentCount();
			//	var slot = (uint)description.Bindings[i].Semantics;

			//	Assert.NotNull(description.Bindings[i].VertexBuffer);

			//	var handle = ((BufferGL3)description.Bindings[i].VertexBuffer.BufferObject).Handle;
			//	_gl.BindBuffer(GL.ArrayBuffer, handle);
			//	_gl.EnableVertexAttribArray(slot);

			//	var normalize = description.Bindings[i].Semantics == DataSemantics.Color0 ||
			//					description.Bindings[i].Semantics == DataSemantics.Color1 ||
			//					description.Bindings[i].Semantics == DataSemantics.Color2 ||
			//					description.Bindings[i].Semantics == DataSemantics.Color3;

			//	_gl.VertexAttribPointer(slot, size, type, normalize, description.Bindings[i].Stride, (void*)description.Bindings[i].Offset);
			//	_gl.VertexAttribDivisor(slot, description.Bindings[i].StepRate);
			//}

			//_gl.BindVertexArray(0);
		}

		/// <summary>
		///     Gets the native vertex layout.
		/// </summary>
		internal uint NativeObject { get; private set; }

		/// <summary>
		///     Gets the offset into the index buffer.
		/// </summary>
		internal int IndexOffset { get; private set; }

		/// <summary>
		///     Gets the type of the indices.
		/// </summary>
		internal uint IndexType { get; private set; }

		/// <summary>
		///     Gets the size of the indices in bytes.
		/// </summary>
		internal int IndexSizeInBytes { get; private set; }

		/// <summary>
		///     Binds the input layout to the input-assembler stage of the pipeline.
		/// </summary>
		public void Bind()
		{
			// Do not actually bind the input layout here, as that causes all sorts of problems with buffer updates between
			// the binding of the input layout and the actual draw call using the input layout
			DeviceState.VertexLayout = this;
		}

		/// <summary>
		///     Disposes the object, releasing all managed and unmanaged resources.
		/// </summary>
		protected override void OnDisposing()
		{
			GLContext.Deallocate(_gl.DeleteVertexArrays, NativeObject);
		}
	}
}

File: Source/Pegasus/Platform/Graphics/OpenGL3/VertexLayoutGL3.cpp
#include "Prelude.hpp"

using namespace OpenGL3;

VertexLayout* GraphicsDevice::InitializeVertexLayout(VertexLayoutDescription* desc)
{
	auto vertexLayout = PG_NEW(VertexLayout);
	vertexLayout->Obj = Allocate(&GraphicsDevice::glGenVertexArrays, "Vertex Input Layout");

	glBindVertexArray(vertexLayout->Obj);

	if (desc->IndexBuffer != nullptr)
	{
		glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, static_cast<Buffer*>(desc->IndexBuffer)->Obj);

		vertexLayout->IndexOffset = safe_static_cast<uint32>(desc->IndexOffset);
		vertexLayout->IndexType = Map(desc->IndexSize);
		vertexLayout->IndexSize = safe_static_cast<uint32>(GetIndexSizeInBytes(desc->IndexSize));
	}

	for (auto i = 0; i < desc->BindingsCount; ++i)
	{
		auto type = GetVertexDataType(desc->Bindings[i].Format);
		auto size = GetVertexDataComponentCount(desc->Bindings[i].Format);
		auto slot = static_cast<uint32>(desc->Bindings[i].Semantics);

		PG_ASSERT_NOT_NULL(desc->Bindings[i].VertexBuffer);

		glBindBuffer(GL_ARRAY_BUFFER, static_cast<Buffer*>(desc->Bindings[i].VertexBuffer)->Obj);
		glEnableVertexAttribArray(slot);

		// Color values must be normalized
		auto normalize = desc->Bindings[i].Semantics == DataSemantics::Color0 ||
			desc->Bindings[i].Semantics == DataSemantics::Color1 ||
			desc->Bindings[i].Semantics == DataSemantics::Color2 ||
			desc->Bindings[i].Semantics == DataSemantics::Color3;

		glVertexAttribPointer(slot, size, type, static_cast<int32>(normalize), desc->Bindings[i].Stride, ToPointer(desc->Bindings[i].Offset));
		glVertexAttribDivisor(slot, desc->Bindings[i].StepRate);
	}

	glBindVertexArray(0);
	return vertexLayout;
}

void GraphicsDevice::FreeVertexLayout(VertexLayout* vertexLayout)
{
	if (vertexLayout == nullptr)
		return;

	glDeleteVertexArrays(1, &vertexLayout->Obj);
	PG_DELETE(vertexLayout);
}

void GraphicsDevice::BindVertexLayout(VertexLayout* vertexLayout)
{
	// Do not actually bind the input layout here, as that causes all sorts of problems with buffer updates between
	// the binding of the input layout and the actual draw call using the input layout
	_boundVertexLayout = vertexLayout;
}